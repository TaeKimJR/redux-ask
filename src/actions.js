import { createAction } from 'redux-actions';
import { createInfiniteAction } from 'infinite-reducer';
import { REQUEST_REDUCER_IDENTIFIER } from './constants';
import * as types from './types';

const setInitialUnwrapped = createAction(types.SET_INITIAL);
const setPendingUnwrapped = createAction(types.SET_PENDING);
const setSuccessUnwrapped = createAction(types.SET_SUCCESS);
const setFailureUnwrapped = createAction(types.SET_FAILURE);

export const unwrapped = {
  setInitial: setInitialUnwrapped,
  setPending: setPendingUnwrapped,
  setSuccess: setSuccessUnwrapped,
  setFailure: setFailureUnwrapped,
};

const setInitial = createInfiniteAction(REQUEST_REDUCER_IDENTIFIER, setInitialUnwrapped);
const setPending = createInfiniteAction(REQUEST_REDUCER_IDENTIFIER, setPendingUnwrapped);
const setSuccess = createInfiniteAction(REQUEST_REDUCER_IDENTIFIER, setSuccessUnwrapped);
const setFailure = createInfiniteAction(REQUEST_REDUCER_IDENTIFIER, setFailureUnwrapped);

export {
  setInitial,
  setPending,
  setSuccess,
  setFailure,
};
