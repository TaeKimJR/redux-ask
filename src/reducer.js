import * as R from 'ramda';
import { handleActions } from 'redux-actions';
import { combineReducers } from 'redux';
import { createInfiniteReducer } from 'infinite-reducer';
import { REQUEST_REDUCER_IDENTIFIER, status } from './constants';
import * as types from './types';

// Used for each individual requests
const requestInitialState = {
  status: status.NOT_STARTED,
  response: null,
  error: null,
};
export const requestReducer = handleActions(
  {
    [types.SET_INITIAL]: R.always(requestInitialState),
    [types.SET_PENDING]: state => R.assoc('status', status.PENDING, state),
    [types.SET_SUCCESS]: (state, { payload }) => R.compose(
      R.assoc('response', payload),
      R.assoc('status', status.SUCCESS),
    )(state),
    [types.SET_FAILURE]: (state, { payload }) => R.compose(
      R.assoc('error', payload),
      R.assoc('status', status.FAILURE),
    )(state),
  },
  requestInitialState,
);

// Used for all requests, sharing the requestReducer
const requestReducers = createInfiniteReducer(REQUEST_REDUCER_IDENTIFIER, requestReducer);


// Used for configuring the request api
const configInitialState = {
  options: {},
  onUnauthenticated: null,
  onSuccess: null,
  onFailure: null,
};

export const configReducer = handleActions(
  {
    [types.SET_CONFIG]: (state, { payload }) => R.merge(configInitialState, payload),
  },
  configInitialState,
);

// Used to create the reducer (allowing for configuration overrides passed in)
const reducer = combineReducers({
  config: configReducer,
  requests: requestReducers,
});

export default reducer;

