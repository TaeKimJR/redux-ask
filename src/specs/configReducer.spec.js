import { configReducer } from '../reducer';
import * as actions from '../actions';

describe('[Requests] Config Reducer', () => {
  describe('Initial State', () => {
    test('should use the default empty config when no initial state is given', () => {
      const reduced = configReducer(undefined, {});

      expect(reduced).toMatchSnapshot();
    });

    test('should override any config with state given', () => {
      const overrides = {
        options: { foo: 'bar' },
        onFailure: () => {},
      };

      const reduced = configReducer(overrides, {});

      expect(reduced).toMatchSnapshot();
    });
  });

  test('SET_CONFIG', () => {
    const config = {
      options: {
        headers: {
          'Content-type': 'application/json',
        },
      },
      onUnauthenticated: () => () => 'unauthenticated',
      onSuccess: () => () => 'success',
      onFailure: () => () => 'failure',
    };
    const state = {};

    const reduced = configReducer(state, actions.setRequestConfig(config));

    expect(reduced).toMatchSnapshot();
  });
});
