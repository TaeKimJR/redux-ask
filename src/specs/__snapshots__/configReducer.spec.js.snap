// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Requests] Config Reducer Initial State should override any config with state given 1`] = `
Object {
  "onFailure": [Function],
  "options": Object {
    "foo": "bar",
  },
}
`;

exports[`[Requests] Config Reducer Initial State should use the default empty config when no initial state is given 1`] = `
Object {
  "onFailure": null,
  "onSuccess": null,
  "onUnauthenticated": null,
  "options": Object {},
}
`;

exports[`[Requests] Config Reducer SET_CONFIG 1`] = `
Object {
  "onFailure": [Function],
  "onSuccess": [Function],
  "onUnauthenticated": [Function],
  "options": Object {
    "headers": Object {
      "Content-type": "application/json",
    },
  },
}
`;
